snuba/datasets/plans/query_plan.py

```diff
@@ -2,43 +2,108 @@
-    contrarily to the DB Query Processors, still provided by the storage, which
+    Provides the directions to execute a Clickhouse Query against one
+    storage or multiple joined ones.
+
+    This is produced by a QueryPlanner class (provided either by the
+    dataset or by an entity) after the dataset query processing has
+    been performed and the storage/s has/have been selected.
+
+    It embeds the Clickhouse Query (the query to run on the storage
+    after translation). It also provides a plan execution strategy
+    that takes care of coordinating the execution of the query against
+    the database. The execution strategy can also decide to split the
+    query into multiple chunks.
+
+    TODO: Bring the methods that apply the processors in the plan itself.
+    Now that we have two Plan implementations with a different
+    structure, all code depending on this would have to be written
+    differently depending on the implementation.
+    Having the methods inside the plan would make this simpler.
+    """
+
+    query: TQuery
+    execution_strategy: QueryPlanExecutionStrategy[TQuery]
+
+
+@dataclass(frozen=True)
+class ClickhouseQueryPlan(QueryPlan[Query]):
+    """
+    Query plan for a single entity, single storage query.
+
+    It provides the sequence of storage specific QueryProcessors
+    to apply to the query after the the storage has been selected.
+    These are divided in two sequences: plan processors and DB
+    processors.
+    Plan processors have to be executed only once per plan contrarily
+    to the DB Query Processors, still provided by the storage, which
     are executed by the execution strategy at every DB Query.
-    It also provides a plan execution strategy that takes care of coordinating the
-    execution of the query against the database. The execution strategy can also decide
-    to split the query into multiple chunks.
     """
 
-    query: Query
     plan_query_processors: Sequence[QueryProcessor]
     db_query_processors: Sequence[QueryProcessor]
-    execution_strategy: QueryPlanExecutionStrategy
 
 
-class QueryPlanExecutionStrategy(ABC):
+@dataclass(frozen=True)
+class SubqueryProcessors:
+    """
+    Collects the query processors to execute on each subquery
+    in a composite query.
+    """
+
+    plan_processors: Sequence[QueryProcessor]
+    db_processors: Sequence[QueryProcessor]
+
+
+@dataclass(frozen=True)
+class CompositeQueryPlan(QueryPlan[CompositeQuery[Table]]):
+    """
+    Query plan for the execution of a Composite Query.
+
+    This contains the composite query, the composite strategy and
+    the processors for all the simple subqueries.
+    SubqueryProcessors instances contain all the storage query
+    processors that still have to be executed on the subquery.
+    """
+
+    # If there is no join there would be no table alias and one
+    # single simple subquery.
+    root_processors: Optional[SubqueryProcessors]
```

Comment:
When would there be a composite query with a single subquery?

Reply:
```
SELECT avg(soemthing) FROM (
     SELECT uniq(release) FROM sentry ..... GROUP BY platform
)
```

Reply:
Oh yeah of course.